lib/ansible/modules/windows/win_nssm.ps1

```diff
@@ -34,31 +36,21 @@ if (($appParameters -ne $null) -and ($appParametersFree -ne $null))
 {
     Fail-Json $result "Use either app_parameters or app_parameteres_free_form, but not both"
 }
+if (($appParameters -ne $null) -and ($appParameters -isnot [System.Collections.Hashtable])) {
```

Comment:
As much as I would like to force it being a dict/hashtable we still need to support the older option (string representing a dict). We should be adding a deprecation warning saying that this format will be removed in 4 releases. In reality we should remove the whole dict/hashtable part as well but see my comment above.

Reply:
I saw a similar comment on issue https://github.com/ansible/ansible/issues/25265#issuecomment-306271563, but I don't understand what is it exactly.

The module documentation doesn't give any example of this usage. Is it really something that is supposed to be supported ? Where does it come from ?

That said, from my experience, currently the app_parameters parameter doesn't work at all on any of the supported Ansible versions (see #25265), so maybe it's not a big deal to break a little while fixing almost everything else, as a first step before doing what is said in the comment above...

Reply:
I'm honestly not sure, I need to play around with it tomorrow and have a closer look at the code. Weird things are happening in this module so I'll get back to you tomorrow about this.

Reply:
Ok, sorry @ksubileau for taking so long but here is what I found out

* The docs for win_nssm gives examples on specifying `app_parameters` as a literal YMAL dict which worked in previous versions of Ansible
* The commit https://github.com/ansible/ansible/commit/ff1efb2c1c2cbaebe3ed28ca28695b9bc15d5bca#diff-5ba1945628b98dfd2ff1c1893772b39a added `-type "str"` to the value which meant a the value would have been converted using `.ToString()`, in this case a hashtable ToString value is `System.Collections.Hashtable`
* The commit above effective broke compatibility with someone specifying a literal dict but `app_parameters` still supports the `key=value` format
* This was added in the 2.3.0 release so this example has been broken since then

What I think we should do is;

* Update the existing docs in a separate PR to show that examples using the `key=value` syntax
* Don't re-add support for specifying a dict, we've already ripped the bandaid off so there's no sense in us going through the pain once again if we planned on deprecating it
* Deprecate `app_parameters` saying it will be removed in Ansible 2.12 as supporting a key=value syntax for arguments is a bit silly
* Alias `app_parameters_free_form` to a new option called `arguments` of the type `str`

I know you said the name `AppParameters` is used internally in nssm but I think we should still move away from that and use something that is a bit more standard in Ansible. My reasoning for this are;

* The gui for editing an nssm service has the relevant field as `Arguments`
* The install command line docs at [nssm](https://nssm.cc/commands) show the value as `<arguments>`
* There is a section around add/change/querying individual parameters that align with the `AppParameters` name but Ansible is more a state driven engine than an imperative action on individual args, e.g. a definition for win_nssm is more about specifying all the args for the service and not removing a particular one
* The screenshot you sent through was for the registry so it's more an internal naming standard in nssm and not necessarily what they expose to end users

Hopefully this makes sense but ultimately it means we can get rid of the Hashtable handling code and cordon off the key=value until 2.12 where we can then remove it.

Reply:
Ok, I understand your reasoning but indeed I made the opposite reasoning while doing this PR. For me, what is specified in the module's current documentation is what should work. So I fixed the code so that the YAML dict notation works, and I tried to make only minimal changes so that this PR could be backported easily.
In the other side, AFAIK the key=value syntax for this option has never been documented, so I think we can assume that very few users use it.

So, taking into account your previous comment, here is my proposal: 
 * In this bugfix PR, for the upcoming 2.7 and possibly backported in 2.6:
   * Restore support of the  `key=value` format (broken with the current code of this PR).
   * Keep support for specifying a YAML dict. Broken since 2.3.0, but that's what's specified in the doc, the work is already done, and the code will be shared with the `key=value` format. If this PR is backported, it will work as specified in the doc for versions 2.6.x and 2.7
   * Eventually update the doc to show examples using the `key=value` syntax (here or in a separate PR)

 * In a next feature PR, for 2.8 or another future release:
   * Deprecate `app_parameters`
   * Alias `app_parameters_free_form` to a new option called `arguments`
   * Add support of list form to `arguments`
   * Maybe also add check mode, add more NSSM options like AppDirectory, and more...

I've already started working on this next PR, and the first three points are already done (see https://github.com/ksubileau/ansible/commit/20cc3b972c06312bdc775894a6a2c03490c51130)

Let me know if you agree with this plan, I think this may be a good compromise ;)

Reply:
No worries, I totally understand your reasoning with adding back the dict support but because we have had 4 (soon to be 5) releases where the YAML dict format has not worked I don't see the point in adding it back in considering no playbook since 2.2 would have worked.

Are you on IRC, it would be great to talk to you in an easier fashion and we can sum up what to do in one post once we've come to an agreement? My nickname on there is jborean93 but we can also just chat on the #ansible-windows channel.